feat: login notification email

a-lider opened the pull request:
## Problem
Users need to be notified when someone logs into their PostHog account for security purposes. Email should be sent only on login from a new device, not on signup or re-authentication. 

Closes #35299 

## Changes

- Added Customer.io and SMTP templates for login notifications.  
- Extended the existing `post_login` signal receiver to detect new device logins and trigger login notifications.  
- Implemented `check_and_cache_login_device()` to store device fingerprints (IP + user agent hash) in Redis for 30 days to avoid sending notifications when the user logs in from the same device.

<img width="1369" height="735" alt="Screenshot 2025-07-28 at 12 59 51" src="https://github.com/user-attachments/assets/d882c64f-5cc8-47d1-a7f7-372eea23c959" />

## How did you test this code?

- Created new account - verified no login notification sent
- Logged in from new browser - verified notification email received with device info
  - Note: when running locally, ip_address is localhost and GeoIP data is unavailable, so location is Unknown. 
- Verified same device login doesn't trigger notification (Redis cache)
- Verified quick re-auth doesn't send email

Commit:
feat: new device login notification email

Commit:
Do not send login notification on signup

yasen-posthog commented on posthog/api/authentication.py:
```diff
@@ -39,17 +41,71 @@
+                browser=browser_info,
+            )
+        else:
+            capture_exception(Exception("GeoIP lookup failed"), additional_properties={"ip_address": ip_address})
```
Wouldn't that fire each time the user logs in, has GeoIP data but the device is not new?

zlwaterfield reviewed (changes requested):
Awesome work here, feels very close. Just some things to iterate on.

yasen-posthog commented on posthog/api/authentication.py:
```diff
@@ -39,17 +41,71 @@
+from posthog.geoip import get_geoip_properties
+
+
+def check_and_cache_login_device(user_id: int, ip_address: str, user_agent: str) -> bool:
```
+1

Commit:
Refactor login notification

- Move device caching to posthog/caching
- Move new device detection to Celery task

Commit:
Add unit tests

zlwaterfield commented on posthog/caching/login_device_cache.py:
```diff
@@ -0,0 +1,24 @@
+from posthog.redis import get_client
+
+
+def check_and_cache_login_device(user_id: int, ip_address: str, short_user_agent: str) -> bool:
```
We should also add tests for this if you're able to figure out the redis tests, should be lots of examples.

zlwaterfield reviewed (commented):
Added two more comments, after that it looks great

Commit:
Capture login notification sent

Commit:
test: login device caching

Commit:
test: send login notification email

a-lider committed:
Merge branch 'master' into feat/login-notification

Commit:
fix: Django Tests CI Error

Commit:
Keep just the simple device caching test

Commit:
Revert Redis tests

Commit:
Check if email is available

Commit:
test: login device caching

zlwaterfield committed:
Merge branch 'master' into feat/login-notification

Commit:
fix: send login notification only for cloud deployment

a-lider committed:
Merge branch 'master' into feat/login-notification

Commit:
Fix: Clean up Redis keys after each test

Commit:
Fix: Clean up Redis keys before each test

Commit:
Fix: do not send login notification in test env, it breaks other tests

Commit:
Fix: do not send login notification in cli

Commit:
Fix typo

Commit:
fix: send login notification only for cloud deployment

Commit:
Clean up all keys in Redis before each test

Commit:
fix: avoid potential user_id collision

Commit:
fix: check settings.TEST

zlwaterfield committed:
Update expected query count in middleware test

Adjusted the expected number of database queries in the 'test_project_switched_when_accessing_dashboard_of_another_accessible_team' test from 7 to 8 to reflect recent changes in query behavior.

zlwaterfield commented:
@a-lider - one note: might be worth adding a warning to the email for the first few weeks saying something like: "Seeing this for the first time? This is a new email that we're sending to help improve the security on the PostHog platform". Just because it might be weird to see that if they login from the same device they always login from and get it.

Commit:
Test: check login notification email content

a-lider committed:
Merge branch 'master' into feat/login-notification

Commit:
fix: expected query count in middleware test

zlwaterfield reviewed (approved):
Woo!
